File: firstapp/samples/pkgcloud/getting_started.js
// step-1

auth_username = 'your_auth_username';
auth_password = 'your_auth_password';
auth_url = 'http://controller:5000';
project_name = 'your_project_name_or_id';
region_name = 'your_region_name';


var conn = require('pkgcloud').compute.createClient({
    provider: 'openstack',
    username: auth_username,
    password: auth_password,
    authUrl:auth_url,
    region: region_name
});


// step-2
conn.getImages(function(err, images) {
    for (i =0; i<images.length; i++) {
        console.log("id: " + images[i].id);
        console.log("name: " + images[i].name);
        console.log("created: " + images[i].created);
        console.log("updated: " + images[i].updated);
        console.log("status: " + images[i].status + "\n");
}});


// step-3
conn.getFlavors(function(err, flavors) {
    for (i=0; i<flavors.length; i++) {
        console.log("id: " + flavors[i].id);
        console.log("name: " + flavors[i].name);
        console.log("ram: " + flavors[i].ram);
        console.log("disk: " + flavors[i].disk);
        console.log("vcpus: " + flavors[i].vcpus + "\n");
}});


// step-4
image_id = '2cccbea0-cea9-4f86-a3ed-065c652adda5';
conn.getImage(image_id, function(err, image) {
    console.log("id: " + image.id);
    console.log("name: " + image.name);
    console.log("created: " + image.created);
    console.log("updated: " + image.updated);
    console.log("status: " + image.status + "\n");
});


// step-5
flavor_id = '3';
conn.getFlavor(flavor_id, function(err, flavor) {
        console.log("id: " + flavor.id);
        console.log("name: " + flavor.name);
        console.log("ram: " + flavor.ram);
        console.log("disk: " + flavor.disk);
        console.log("vcpus: " + flavor.vcpus + "\n");
});


// step-6
instance_name = 'testing';
conn.createServer({
  name: instance_name,
  flavor: flavor_id,
  image: image_id
}, function(err, server) {
    console.log(server.id)
});


// step-7
conn.getServers(console.log);
// TODO - make a decision about printing this nicely or not

// step-8
test_instance_id = '0d7968dc-4bf4-4e01-b822-43c9c1080d77';
conn.destroyServer(test_instance_id, console.log);
// TODO - make a decision about printing this nicely or not

// step-9
console.log('Checking for existing SSH key pair...');
keypair_name = 'demokey';
pub_key_file = '/home/user/.ssh/id_rsa.pub';
pub_key_string = '';
keypair_exists = false;
conn.listKeys(function (err, keys) {
    for (i=0; i<keys.length; i++){
        if (keys[i].keypair.name == keypair_name) {
            keypair_exists = true;
}}});


if (keypair_exists) {
    console.log('Keypair already exists.  Skipping import.');
} else {
    console.log('adding keypair...');
    fs = require('fs');
    fs.readFile(pub_key_file, 'utf8', function (err, data) {
      pub_key_string = data;
    });
    conn.addKey({name: keypair_name, public_key:pub_key_string}, console.log);
}

conn.listKeys(function (err, keys) {
    for (i=0; i<keys.length; i++){
        console.log(keys[i].keypair.name)
        console.log(keys[i].keypair.fingerprint)
}});

// step-10
security_group_name = 'all-in-one';
security_group_exists = false;
all_in_one_security_group = false;
conn.listGroups(function (err, groups) {
    for (i=0; i<groups.length; i++){
        if (groups[i].name == security_group_name) {
            security_group_exists = true;
}}});

if (security_group_exists) {
    console.log('Security Group already exists.  Skipping creation.');
} else {
    conn.addGroup({ name: 'all-in-one',
                    description: 'network access for all-in-one application.'
                  }, function (err, group) {
    all_in_one_security_group = group.id;
    conn.addRule({ groupId: group.id,
                   ipProtocol: 'TCP',
                   fromPort: 80,
                   toPort: 80}, console.log);
    conn.addRule({ groupId: group.id,
                   ipProtocol: 'TCP',
                   fromPort: 22,
                   toPort: 22}, console.log);
   });
};

// step-11
userdata = "#!/usr/bin/env bash\n" +
    "curl -L -s https://git.openstack.org/cgit/stackforge/faafo/plain/contrib/install.sh" +
    " | bash -s -- -i faafo -i messaging -r api -r worker -r demo";
userdata = new Buffer(userdata).toString('base64')

// step-12
instance_name = 'all-in-one'
conn.createServer({ name: instance_name,
                    image: image_id,
                    flavor: flavor_id,
                    keyname: keypair_name,
                    cloudConfig: userdata,
                    securityGroups: all_in_one_security_group},
                    function(err, server) {
                        server.setWait({ status: server.STATUS.running }, 5000, console.log)
                    });

// step-13
console.log('Checking for unused Floating IP...')
unused_floating_ips = []
conn.getFloatingIps(function (err, ips) {
    console.log(ips)
    for (i=0; i<ips.length; i++){
        if (ips[i].node_id) {
            unused_floating_ips = ips[i];
            break;
}}});


if (!unused_floating_ip) {
    conn.allocateNewFloatingIp(function (err, ip) {
    unused_floating_ip = ip.ip;
})};

console.log(unused_floating_ip);

// step-14
conn.addFloatingIp(testing_instance, unused_floating_ip, console.log)

// step-15
console.log('The Fractals app will be deployed to http://%s' % unused_floating_ip.ip_address)

File: firstapp/samples/shade/getting_started.py
#step-1
from shade import *

simple_logging(debug=True)
conn = openstack_cloud(cloud='myfavoriteopenstack')

#step-2
images = conn.list_images()
for image in images:
    print(image)

#step-3
flavors =  conn.list_flavors()
for flavor in flavors:
    print(flavor)

#step-4
image_id = 'c55094e9-699c-4da9-95b4-2e2e75f4c66e'
image = conn.get_image(image_id)
print(image)

#step-5
flavor_id = '100'
flavor = conn.get_flavor(flavor_id)
print(flavor)

#step-6
instance_name = 'testing'
testing_instance = conn.create_server(wait=True, auto_ip=True,
    name=instance_name,
    image=image_id,
    flavor=flavor_id)
print(testing_instance)

#step-7
instances = conn.list_servers()
for instance in instances:
    print(instance)

#step-8
conn.delete_server(name=instance_name)

#step-9
print('Checking for existing SSH keypair...')
keypair_name = 'demokey'
pub_key_file = '~/.ssh/demo_key.pub'

if conn.search_keypairs(keypair_name):
    print('Keypair already exists. Skipping import.')
else:
    print('Adding keypair...')
    conn.create_keypair(keypair_name, open(pub_key_file, 'r').read().strip())

for keypair in conn.list_keypairs():
    print(keypair)

#step-10
print('Checking for existing security groups...')
sec_group_name = 'all-in-one'
if conn.search_security_groups(sec_group_name):
    print('Security group already exists. Skipping creation.')
else:
    print('Creating security group.')
    conn.create_security_group(sec_group_name, 'network access for all-in-one application.')
    conn.create_security_group_rule(sec_group_name, 80, 80, 'TCP')
    conn.create_security_group_rule(sec_group_name, 22, 22, 'TCP')

conn.search_security_groups(sec_group_name)

#step-11
ex_userdata = '''#!/usr/bin/env bash

curl -L -s https://git.openstack.org/cgit/openstack/faafo/plain/contrib/install.sh | bash -s -- \
-i faafo -i messaging -r api -r worker -r demo
'''

#step-12
instance_name = 'all-in-one'
testing_instance = conn.create_server(wait=True, auto_ip=False,
    name=instance_name,
    image=image_id,
    flavor=flavor_id,
    key_name=keypair_name,
    security_groups=[sec_group_name],
    userdata=ex_userdata)

#step-13
f_ip = conn.available_floating_ip()

#step-14
conn.add_ip_list(testing_instance, [f_ip['floating_ip_address']])

#step-15
print('The Fractals app will be deployed to http://%s' % f_ip['floating_ip_address'] )
